### PR TITLE
Moved class definitions from impl headers to interface headers as nee…

### PR DESCRIPTION
**Pull request by zajo**
…ded. Consistent application of inline throughout url_base/view sources.

### CODE REVIEW

**Comment by vinniefalco** [comment]
Commit message is too long, "Refactor source files" is better

File: include/boost/url/impl/url_base.hpp [hunk 1]
```diff
@@ -22,194 +22,48 @@ segments_type(
...
-params_type(
-    url_base const& v) noexcept
-    : s_(v.s_)
-    , pt_(&v.pt_)
+bool
+url_base::
+segments_type::
+iterator::
+operator==(
+    iterator const& other) const noexcept
```
Comment (vinniefalco) [inline include/boost/url/impl/url_base.hpp:30]: iterators are passed by value

Comment (zajo) [reply include/boost/url/impl/url_base.hpp:30]: You've done this in many places, you want me to change it?

Comment (zajo) [reply include/boost/url/impl/url_base.hpp]: I changed it. You were only doing this in op== and op!=.

File: include/boost/url/impl/url_view.hpp [hunk 2]
```diff
@@ -13,173 +13,21 @@
...
-
-//----------------------------------------------------------
-
-class url_view::params_type::iterator
+bool
+url_view::
+segments_type::
+iterator::
+operator==(
+    iterator const& other) const noexcept
```
Comment (vinniefalco) [inline include/boost/url/impl/url_view.hpp]: iterators are passed by value

File: include/boost/url/url_base.hpp [hunk 3]
```diff
@@ -1824,6 +1820,92 @@ class url_base::segments_type::value_type
...
+    using value_type =
+        segments_type::value_type;
+
+    /// A pointer to an element
+    using pointer = value_type const*;
+
+    /// A reference to an element
+    using reference = value_type const&;
+
+    /// The difference_type for this iterator
```
Comment (vinniefalco) [inline include/boost/url/url_base.hpp:1849]: Use the brief from JSON, I think it is something about a signed integer type

File: include/boost/url/url_base.hpp [hunk 4]
```diff
@@ -1824,6 +1820,92 @@ class url_base::segments_type::value_type
...
+    BOOST_URL_DECL
+    iterator(
+        url_base* v,
+        bool end) noexcept;
+
+public:
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    using value_type =
```
Comment (vinniefalco) [inline include/boost/url/url_base.hpp:1840]: These need javadocs

File: include/boost/url/url_base.hpp [hunk 5]
```diff
@@ -1824,6 +1820,92 @@ class url_base::segments_type::value_type
...
+        ++*this;
+        return tmp;
+    }
+
+    BOOST_URL_DECL
+    iterator&
+    operator--() noexcept;
+
+    iterator
+    operator--(int) noexcept
```
Comment (vinniefalco) [inline include/boost/url/url_base.hpp:1894]: These all need javadocs

File: include/boost/url/impl/url_base.hpp (hunk 1 above)
Comment (vinniefalco) [reply include/boost/url/impl/url_base.hpp]: If I passed by `const&`, it was wrong

**Comment by vinniefalco** [comment]
Since this is just shuffling code around, ignore my comments about javadocs for this pull request. However they will need to be added in a separate commit in another pull request.

**Comment by vinniefalco** [comment]
The commit message is too verbose:
image

Suggested: "Declarations are in public headers"

**Comment by vinniefalco** [comment]
Never use past tense in commit messages, it is additional unnecessary verbosity

File: include/boost/url/impl/url_base.hpp [hunk 6]
```diff
@@ -22,194 +22,48 @@ segments_type(
...
+iterator::
+operator==(
+    iterator other) const noexcept
 {
+    BOOST_ASSERT(
+        v_ != other.v_ ||
+        off_ != other.off_ ||
+        n_ == other.n_);
+    return
+        v_ == other.v_ &&
```
Comment (vinniefalco) [inline include/boost/url/impl/url_base.hpp:37]: C++ standard leaves comparison of iterators from different containers as undefined. Thus this code is not quite correct. We should assert that the containers are the same. And when we return the value we should not check that the containers are the same.

File: include/boost/url/impl/url_base.hpp [hunk 7]
```diff
@@ -224,22 +78,21 @@ at( string_view key,
...
+    iterator other) const noexcept
 {
-    return segments_type(*this);
+    BOOST_ASSERT(
+        v_ != other.v_ ||
+        off_ != other.off_ || (
+            nk_ == other.nk_ &&
+            nv_ == other.nv_));
+    return
+        v_ == other.v_ &&
```
Comment (vinniefalco) [inline include/boost/url/impl/url_base.hpp:94]: Same here about checking the container

File: include/boost/url/impl/url_base.ipp [hunk 8]
```diff
@@ -1375,7 +1375,8 @@ erase( iterator first, iterator last ) noexcept ->
     std::memmove(v.s_ + first.off_, v.s_ + last.off_, v.pt_.offset[detail::id_end] - last.off_ + 1);
     v.pt_.resize(detail::id_path, v.pt_.length(detail::id_path, detail::id_query) - d);
     BOOST_ASSERT(v.size() + d == v.a_.size());
-    v.a_.resize(v.size());
+    auto const s = v.a_.resize(v.size());
+    BOOST_ASSERT(v.s_ == s);
```
Comment (vinniefalco) [inline include/boost/url/impl/url_base.ipp:1379]: This is unnecessary, you shouldn't be calling resize if the new size is smaller

**Pull request merged**